src/MangaJet/app/src/main/java/com/mangajet/mangajet/ui/search/SearchViewModel.kt

```diff
@@ -1,78 +1,111 @@
+import androidx.fragment.app.FragmentManager
 import androidx.lifecycle.ViewModel
 import com.mangajet.mangajet.MangaJetApp.Companion.context
+import com.mangajet.mangajet.MangaListAdapter
+import com.mangajet.mangajet.MangaListElementContainer
 import com.mangajet.mangajet.data.Librarian
 import com.mangajet.mangajet.data.Manga
+import com.mangajet.mangajet.data.MangaJetException
+import com.mangajet.mangajet.databinding.SearchFragmentBinding
+import kotlinx.coroutines.ensureActive
 import kotlinx.coroutines.Job
-import kotlinx.coroutines.withContext
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.GlobalScope
 import kotlinx.coroutines.launch
-import com.mangajet.mangajet.R
-import com.mangajet.mangajet.data.MangaJetException
-import com.mangajet.mangajet.databinding.SearchFragmentBinding
+import kotlinx.coroutines.withContext
 
 // Class which represents "Search" ViewModel
 class SearchViewModel : ViewModel() {
     companion object {
         const val SEARCH_AMOUNT = 20        // Amount of searchable mangas
+        const val WAIT_FOR_BUILD_IN_MS = 75
     }
-
-
-    val mangasNames = ArrayList<String>()           // mangas names for list
     var mangas : ArrayList<Manga> = arrayListOf()   // mangas for "AboutManga" activity
     var job : Job? = null                           // Async job for searching and uploading
-    var adapter : ArrayAdapter<String>? = null      // adapter for list
+    var adapter : MangaListAdapter? = null          // adapter for list
+
+    // mangas info for list
+    val mangasInfos = ArrayList<MangaListElementContainer>()
+
+    // Selected resources
+    var allLibraries = arrayOf(
+        Librarian.LibraryName.Readmanga,
+        Librarian.LibraryName.Mangachan,
+        Librarian.LibraryName.Mangalib,
+        Librarian.LibraryName.Acomics
+    )
+    // flags for each resource
+    private var chosenLibraries = BooleanArray(allLibraries.size)
+
+    // init flags
+    init {
+        chosenLibraries[0] = true
+        for (i in 1 until chosenLibraries.size)
+            chosenLibraries[i] = false
+    }
 
     // Function which will upload manga into mangas array and catch exceptions
-    private suspend fun uploadMangaIntoArray(i : Int) {
+    private suspend fun uploadMangaIntoArray(i : Int) : Boolean {
         try {
+            job?.ensureActive()
             mangas[i].updateInfo()
             withContext(Dispatchers.Main) {
-                mangasNames.add(mangas[i].originalName)
+                mangasInfos.add(MangaListElementContainer(
+                    mangas[i].originalName,
+                    mangas[i].author,
+                    mangas[i].library.getURL(),
+                    mangas[i].cover
+                    ))
                 adapter?.notifyDataSetChanged()
             }
         } catch (ex : MangaJetException) {
             withContext(Dispatchers.Main) {
                 Toast.makeText(context, ex.message, Toast.LENGTH_LONG).show()
```

Comment:
To be truth, I do not think that we should Toast every not successful Search element. 
I think in the next sprint we can add logging after you start catching all exceptions, but for now it will be better if we just say nothing to user